### PR TITLE
sql: Rename LogicalColumnID to PGAttributeNum

### LINKED ISSUE #52379: sql: LogicalColumnID is mis-designed
LogicalColumnID was made due to the fact that alter column type recreates a column bumping the column id of the column. Notably, the column id is used to populate values such as `attnum` in pg_catalog.pg_attribute. The column id is also used as the ordinal_position of a column in info_schema-columns https://www.postgresql.org/docs/9.5/infoschema-columns.html.

We want to "column id" to stay consistent after the alter column type operation, thus LogicalColumnID was introduced. However in it's current state, LogicalColumnID can be confusing, discussion here https://github.com/cockroachdb/cockroach/pull/52309. 

It can be unclear which ID to refer to. Currently, it seems only catalog tables want to refer to LogicalColumnID. 

At the very least, ColumnID and LogicalColumnID should be made different types. Also LogicalColumnID may not be the clearest name, maybe OrdinalPosition would be better?

Note: the only operation that changes the column id of a table is alter column type. It is unclear if we can perform alter column type while keeping the column id consistent.

### PR DESCRIPTION
**Pull request by RichardJCai**
It was unclear when LogicalColumnID should be used.
PGAttributeNum makes it clear that it should only be
referenced it catalog tables.

Fixes https://github.com/cockroachdb/cockroach/issues/52379

Release note: None

### CODE REVIEW

File: pkg/sql/plan_opt.go [hunk 1]
```diff
@@ -143,7 +143,13 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error)
 			// Convert the metadata opt.ColumnID to its ordinal position in the table.
 			colOrdinal := colMeta.Table.ColumnOrdinal(col.ID)
 			// Use that ordinal position to retrieve the column's stable ID.
-			resultCols[i].PGAttributeNum = sqlbase.ColumnID(tab.Column(colOrdinal).ColID())
+			switch tab.Column(colOrdinal).(type) {
```
Comment (rohany) [inline pkg/sql/plan_opt.go:146]: do `t := tab.Column...(.type)` and you dont' have to recast below

File: pkg/sql/sqlbase/structured.proto [hunk 2]
```diff
@@ -129,14 +129,14 @@ message ColumnDescriptor {
...
+  // PGAttributeNum must be accessed through the accessor, since it is set
   // lazily, it is incorrect to access it directly.
-  // LogicalColumnID represents a column's number in catalog tables.
+  // PGAttributeNum represents a column's number in catalog tables.
   // This only differs from ID when the Column order is swapped or
   // the ColumnDescriptor must be remade while remaining visual ordering.
   // This does not exist in TableDescriptors pre 20.2.
-  optional uint32 logical_id = 13 [(gogoproto.nullable) = false,
-   (gogoproto.customname) = "LogicalColumnID", (gogoproto.casttype) = "ColumnID"];
+  optional uint32 pg_attribute_num = 13 [(gogoproto.nullable) = false,
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.proto:138]: nit: just move this to one line

**Review by rohany — APPROVED** [review]

**Comment by RichardJCai** [comment]
bors r=rohany

**Pull request merged**